client/src/components/VideoCall.jsx

```diff
@@ -1,48 +1,165 @@
-                    
-                }}
-            />
-        </div>
-    )
-
-}
+import React, { useEffect, useState } from 'react';
+import eyeson, { StreamHelpers } from 'eyeson';
+import Video from './Video';
+import './VideoCall.css';
+import { usePlayer, useStage } from '@empirica/player';
+
+export function VideoCall ({ roomName, record }) {
+  const [prepared, setPrepared] = useState(false);
+  const [started, setStarted] = useState(false);
+  const [localStream, setLocalStream] = useState(null);
+  const [remoteStream, setRemoteStream] = useState(null);
+  const [audio, setAudio] = useState(true);
+  const [video, setVideo] = useState(true);
+
+  const player = usePlayer();
+  const stage = useStage();
+
+  const handleEvent = event => {
+    const { type } = event;
+    console.debug(type, event);
+    if (type === 'room_setup') {
+      if (record && !event.recording) {
+        eyeson.send({ type: 'start_recording' })
+      }
+      return;
+    }
+    if (type === 'accept') {
+      setLocalStream(event.localStream);
+      setRemoteStream(event.remoteStream);
+      player.set('audioEnabled', true);
+      player.set('videoEnabled', true);
+      // temporary solution to not receiving room_setup event for some reason
+      if (record) {
+        eyeson.send({type: 'start_recording'});
+      }      
+      return;
+    }
+    if (type === 'recording_update') {
+      if (!stage.get('recording_url') && event.recording) {
+        stage.set('recording_url', event.recording.links.self);
+      }
+      return;
+    }
+    if (type === 'stream_update') {
+      setLocalStream(event.localStream);
+      setRemoteStream(event.stream);
+    }
+    if (type === 'warning') {
+      console.log('Warning: ' + event.name);
+      return;
+    }
+    if (type === 'error') {
+      console.log('Error: ' + event.name);
+      endSession();
+      return;
+    }
+    if (type === 'exit') {
+      console.log('Meeting has ended');
+      endSession();
+      return;
+    }
+    console.debug('[App]', 'Ignore received event:', event.type);
+  };
+
+  async function delay(ms) {
+    return new Promise(resolve => {
+      setTimeout(resolve, ms);
+    });
+  }
+
+  const startSession = async () => {
+    let access_key = player.get('accessKey');
+    if (!access_key) {
+      await delay(3000);
+      setPrepared(true);
+      return;
+    }
+    setPrepared(true);
+    setStarted(true);
+    eyeson.onEvent(handleEvent);
+    eyeson.start(access_key);
+  }
+
+  const endSession = () => {
+    player.set('accessKey', null);
+    eyeson.offEvent(handleEvent);
+    eyeson.destroy();
+    eyeson.onEvent(handleEvent);
+    setLocalStream(null);
+    setRemoteStream(null);
+  }
+
+  useEffect(() => {
+    if (roomName && !remoteStream) {
+      startSession();
+    }
+    return () => { 
```

Comment:
if the user refreshes the page, does it close and reopen the session? Is this laggy?

Reply:
unfortunately yes, it will reconnect, which takes 3-5 sec. I don't think there's anyway around this even if I don't put the cleanup. Furthermore, if the cleanup doesn't run, there is a risk that the client's camera will still be on even though it is really recording to nowhere.